Container Groups

shanemcd opened the pull request:
See [`docs/container_groups.md`](https://github.com/shanemcd/awx/blob/toc/docs/container_groups.md) for an overview of this concept.

shanemcd committed:
Allow associating a credential with an instance group

shanemcd committed:
Ignore unison and emacs temporary files

AlanCoding commented on awx/main/models/ha.py:
```diff
@@ -176,6 +176,14 @@ class InstanceGroup(HasPolicyEditsMixin, BaseModel, RelatedJobsMixin):
+        blank=True,
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
```
can this ever be edited by a user? If not, I prefer to set `editable=False`. Also, this should be exposed in the serializer so that people can see what credential is linked via the API. If you just add it to the fields, then it will present itself as a read-only field like this.

shanemcd replied:
@AlanCoding by 'this' do you mean the ability to modify the associated credential? If that's what you meant, then yes they will be able to. [Here](https://tower-mockups.testing.ansible.com/mockups/instance_groups/instance-groups-container-add/) is how this will be surfaced in the UI. InstanceGroups will only have credentials in the context of this new "Container Group" concept. We're still wrestling with the naming here so if you have any thoughts I'd love to hear them.

I should mention that I just copy / pasted this association off of another model, so we should definitely make sure I'm doing the right thing here. 😄 As for exposing through the API - absolutely. I have mostly been working in the task manager up until now but I will be looking to do this soon and may have a question or two for you.

shanemcd committed:
Task manager / scheduler Kubernetes integration

shanemcd committed:
API documentation for container groups

unlikelyzero commented:
@shanemcd , is this PR associated with an awx or tower feature ticket?

jakemcdermott commented on awx/main/models/credential/__init__.py:
```diff
@@ -1092,6 +1096,62 @@ def create(self):
+    namespace='openshift_username_password',
+    kind='kubernetes',
+    name=ugettext_noop('OpenShift Username / Password'),
+    managed_by_tower=True,
```
We shouldnt be passing this keyword argument as it [won't be used](https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L581)

I _do_ see that some of the other MangedCredentialTypes are already doing this as well, but those can be resolved in a different pr / issue as it doesn't directly relate to this PR.

jakemcdermott replied:
We shouldn't be passing this keyword argument as [it won't be used](https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L581).

I do see that some of the other MangedCredentialTypes are already doing this as well, but those can be resolved in a different pr / issue as it doesn't directly relate to this PR.

jakemcdermott commented on awx/main/scheduler/__init__.py:
```diff
@@ -2,3 +2,4 @@
 #
 
 from awx.main.scheduler.task_manager import TaskManager # noqa
```
It looks like we want `TaskManager` and `PodManager` included as part of the public api for `awx.main.scheduler`. You can set the names of these using [the module's `__all__` attribute](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces):

```python
__all__ = ['TaskManager', 'PodManager']
```

You'll see this used in a few other places in awx as well, like [credentials](https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L45). Also, by declaring the names with `__all__` you won't need to disable the lint checks on those import statements.

jakemcdermott commented on awx/main/scheduler/__init__.py:
```diff
@@ -2,3 +2,4 @@
 #
 
 from awx.main.scheduler.task_manager import TaskManager # noqa
+from awx.main.scheduler.kubernetes import PodManager # noqa
```
To avoid tying the package's internal implementation to a specific path, you can use a relative import for these: `from .kubernetes import PodManager`

jakemcdermott commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,74 @@
+
+    def deploy_pod(self):
+        if not self.credential.kubernetes:
+            raise RuntimeError
```
A log or message for this runtime error would be helpful, e.g: 
```python
raise RuntimeError('Pod deployment cannot occur without a Kubernetes credential')
```

jakemcdermott replied:
@shanemcd :wave: just wanted to double-check - did we want to add any sort of message here?

jakemcdermott commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,74 @@
+        self.kube_api.create_namespaced_pod(body=self.pod_definition,
+                                            namespace='awx')
+
+        while True:
```
Can we make it so this loop is _guaranteed_ to end eventually? I see that we're checking against `pod.status`, but maybe some upper limit or retry count would also be good as a fallback?

shanemcd replied:
I ignored this at first with the intention of circling back around and now I am wondering if we need this kind of logic here at all because a) the job itself will hit a timeout if it can't start after a certain amount of time. See `AWX_ISOLATED_LAUNCH_TIMEOUT`. and b) Kubernetes itself will not stay in Pending forever - it will either error out or start. But I hear you. Just not sure if we need it given what I stated. What do you think?

jakemcdermott replied:
That makes sense. It sounds like `AWX_ISOLATED_LAUNCH_TIMEOUT` should have this covered - sounds good to me.

jakemcdermott commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,74 @@
+            pod_spec_override = parse_yaml_or_json(
+                self.task.instance_group.pod_spec_override)
+
+            return {**default_pod_spec, **pod_spec_override}
```
I like this - I think this will be useful and practical from a user perspective. I'm wondering if we can somehow surface the pod spec defaults in the API so that the UI can present them as the default value in the yaml editor, that way users will be able to _see_ the configuration they're extending / overriding.

shanemcd replied:
I am not opposed to the idea at all. Do you have any opinions about how you would like to see that surfaced?

jakemcdermott replied:
I don't have any opinions on how/where to surface these. That said, I think I've seen OPTIONS data used for sending over default values for fields. 

I went and checked our settings endpoints, which do something similar, and they provide default settings in the GET response _and_ in the OPTIONS data:

![Screenshot from 2019-08-12 12-47-51](https://user-images.githubusercontent.com/9753817/62882325-a3a9b000-bcff-11e9-98fb-c2b279c1de8a.png)

:man_shrugging: There may already be a convention for this, but I do not know it. From a UI perspective I think the most important thing would be that nothing needs to be saved or created through the API for the default values to be available.

chrismeyersfsu replied:
Allowing user-provided pod-spec is risky. We would want more assertions around the end result of the pod spec so that we can proceed with certain assumptions in our code that references the podspec. i.e.

https://github.com/ansible/awx/blob/devel/awx/main/fields.py#L418 `json.Draft4Validator(schema).iter_errors` where schema looks like https://github.com/ansible/awx/blob/devel/awx/main/fields.py#L713

shanemcd replied:
@jakemcdermott Implemented here: https://github.com/ansible/awx/pull/4189/commits/4492da22208e4de9cee2ad8def42c9cac957962b#diff-596536a0e5fde93944c40ae984189f0a

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -40,6 +42,51 @@ def __init__(self, cancelled_callback=None, check_callback=None):
+        if credential.get_input('verify_ssl'):
+            ca_cert_data = credential.get_input('ssl_ca_cert')
+            # TODO: Clean up this file after we use it.
+            with tempfile.NamedTemporaryFile(delete=False) as temp:
```
We have a dir that all tmp files for this job should be created. We take care to cleanup these dirs; outside of this task in case of a bad failure.
 https://github.com/ansible/awx/blob/37119a453857b832b3cc9304b8c0f26e4d98e0bc/awx/main/isolated/manager.py#L141

shanemcd replied:
I fixed this up to use `self.private_data_dir`, I think that's what you were talking about? Resolving this convo.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -40,6 +42,51 @@ def __init__(self, cancelled_callback=None, check_callback=None):
+                # where it just doesnt respect the ansible_kubectl_ca_cert var.
+                # This is a workaround.
+                extravars['ansible_kubectl_ca_cert'] = temp.name
+                extravars['ansible_kubectl_extra_args'] = f"--certificate-authority={temp.name}"
```
`f` I need to go google what that does

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -40,6 +42,51 @@ def __init__(self, cancelled_callback=None, check_callback=None):
+
+        return inventory
+
+    @property
```
I don't _love_ that this is a @property . I like to make things a function contain any side effects. Users are encouraged to "call" this property multiple times. If it were a function, a user would be less inclined to do so. In this case, a file is created on disk. Multiple calls to this will result in the ssl ca cert getting written to disk multiple times.

shanemcd replied:
My opinion on this is, from a perspective of application state, this function does not have any side-effects. The fact that it generates a tempfile is irrelevant to the interface and will always return the same output. It does not modify *other* properties of the object, which is the kind of thing I do think we should avoid in a property. That being said, I see your point and I'm on the fence about whether to care or not, so I'll go ahead and give this one to you.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -40,6 +42,51 @@ def __init__(self, cancelled_callback=None, check_callback=None):
+    def kubectl_or_oc_vars(self):
+        extravars = {}
+
+        credential = self.instance.instance_group.credential
```
For the life of me I can't figure out why a singular `instance_group` would hang off of an `instance`. I would expect the plural version. Can you help me out?

chrismeyersfsu replied:
When I run `py.test awx/main/tests/functional/task_management/test_container_groups.py` I get `django.db.utils.OperationalError: no such column: main_instancegroup.credential_id`. Did you maybe miss adding a migration to this PR?

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -393,9 +436,15 @@ def run(self, instance, private_data_dir, playbook, module, module_args,
+
+            if self.instance.is_containerized:
+                ca_cert = self.kubectl_or_oc_vars.get('ansible_kubectl_ca_cert', None)
+                if ca_cert:
```
This cleanup should be removed in favor of a more general cleanup that can account for failure and cleanup.

chrismeyersfsu commented on awx/main/models/credential/__init__.py:
```diff
@@ -136,6 +136,10 @@ def kind(self):
         return self.credential_type.kind == 'cloud'
 
+    @property
+    def kubernetes(self):
```
Would prefer this to be named `is_kubernetes`. Hmm, I can see why you chose the name you chose, to repeat the `cloud` pattern. I _think_ that decision was made so that less code would need to be modified for legacy code reasons.

chrismeyersfsu commented on awx/main/models/credential/__init__.py:
```diff
@@ -1093,6 +1097,38 @@ def create(self):
+            'id': 'ssl_ca_cert',
+            'label': ugettext_noop('Certificate Authority data'),
+            'type': 'string',
+            'secret': True,
```
:+1:

chrismeyersfsu commented on awx/main/models/credential/__init__.py:
```diff
@@ -1093,6 +1097,38 @@ def create(self):
+            'id': 'verify_ssl',
+            'label': ugettext_noop('Verify SSL'),
+            'type': 'boolean',
+            'default': True,
```
:+1:

chrismeyersfsu commented on awx/main/models/ha.py:
```diff
@@ -176,6 +176,18 @@ class InstanceGroup(HasPolicyEditsMixin, BaseModel, RelatedJobsMixin):
+        default=None,
+        on_delete=models.SET_NULL,
+    )
+    pod_spec_override = prevent_search(models.TextField(
```
Nice usage of `prevent_search`

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,96 @@
+    def namespace(self):
+        return self.pod_definition['metadata']['namespace']
+
+    def _kube_api_connection(self, credential):
```
Could make this a static method since it's called in `init()` and doesn't modify the self state.

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,96 @@
+                                                    namespace=self.namespace)
+            if pod.status.phase != 'Pending':
+                break
+            time.sleep(1)
```
We should make a `settings/default.py` for this sleep amount. Maybe a customer can support 500 tower jobs with there super great kubernetes cluster but this could ddos the server.

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -393,9 +436,15 @@ def run(self, instance, private_data_dir, playbook, module, module_args,
             status, rc = self.check()
+
+            if self.instance.is_containerized:
+                ca_cert = self.kubectl_or_oc_vars.get('ansible_kubectl_ca_cert', None)
```
This will call the property, generate a new temp file, then delete that file. You need to get a handle on the original generation of the cert file for the cleanup you intend here. This is why I prefer functions to properties.

chrismeyersfsu commented on awx/main/models/unified_jobs.py:
```diff
@@ -706,6 +706,10 @@ def _get_task_class(cls):
     def supports_isolation(cls):
         return False
 
+    @classmethod
```
Defined as a function here, but when overwritten in jobs.py it is a property.

chrismeyersfsu commented:
I'd rather see the container logic that lives in unified_jobs.py and jobs.py exist as a mixin. Our model files have gotten so bloated it's hard to tell which functions depend on other functions when you are trying to follow a feature.

 i.e.

```
class ContainerizationMixin():
    @property
    def can_run_containerized(self):
        return False

    @property
    def is_containerized(self):
        return False
```

A bunch of concrete examples in our code base https://github.com/ansible/awx/blob/devel/awx/main/models/mixins.py#L406

chrismeyersfsu commented on awx/main/isolated/manager.py:
```diff
@@ -40,6 +42,51 @@ def __init__(self, cancelled_callback=None, check_callback=None):
+        extravars['ansible_kubectl_namespace'] = 'awx'
+
+        if credential.kind == 'kubernetes_bearer_token':
+            extravars['ansible_kubectl_token'] = credential.get_input('bearer_token')
```
I'm not super familiar with our newer credential code. @ryanpetrello Does it matter that `bearer_token` is a secret variable? Specifically, do we have to explicitly decode it .. or does `get_input()` do that for us?

ryanpetrello replied:
`.get_input()` does it for us:

https://github.com/ansible/awx/blob/devel/awx/main/models/credential/__init__.py#L277

chrismeyersfsu commented on awx/main/models/jobs.py:
```diff
@@ -680,6 +680,14 @@ def display_artifacts(self):
+
+    @property
+    def is_containerized(self):
+        return self.instance_group and self.instance_group.is_containerized
```
Would wrap in a bool() cast. If self.instance_group is None and self.instance_group.is_containerized is None then the result return will be None vs. False.

chrismeyersfsu commented on awx/main/models/unified_jobs.py:
```diff
@@ -1407,7 +1411,14 @@ def awx_meta_vars(self):
 
     def get_queue_name(self):
+        if self.is_containerized is True:
+            return settings.CELERY_DEFAULT_QUEUE
```
Can't quite wrap my head around what this should be but something feels wrong. Why shouldn't `self.container_controller` be returned in this case?

chrismeyersfsu replied:
Oh sorry, `container_controller` isn't a model property .. just a temp var I saw earlier in the code. Still, something doesn't feel right here. I think this conditional is not needed and the below return captures all the logic needed.

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,96 @@
+
+        if credential.get_input('verify_ssl'):
+            ca_cert_data = credential.get_input('ssl_ca_cert')
+            configuration.ssl_ca_cert = create_temporary_fifo(ca_cert_data.encode())
```
Hot :+1: But be wary of the thing you are passing this to though. It has to be coded in a way that is pipe friendly.

shanemcd replied:
Yep, this is why I had to use that pesky tempfile on the execution side! Bleh.

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,86 @@
+        self.task = task
+        self.credential = task.instance_group.credential
+        self.kube_api = self._kube_api_connection(self.credential)
+        self.pod_name = f"job-{task.id}"
```
I'm thinking of the case where users have multiple tower servers, dev and prod, pointed at the same kubernetes cluster. Can we add something unique from the tower cluster to the pod name?

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,86 @@
+            }
+        }
+
+        if bool(self.task.instance_group.pod_spec_override):
```
bool() wrapper not needed.

chrismeyersfsu commented:
I'm concerned about getting an error any time we make call(s) to the kubernetes API. It's going to be a he said she said nightmare if the customer cluster is failing or some misconfiguration occurs. Might there be a generic way we can log all kubernetes api errors? Or just have more try excepts sprinkled about?

chrismeyersfsu commented on awx/main/scheduler/task_manager.py:
```diff
@@ -214,13 +215,16 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
                              rampart_group.name, task.log_format))
                 return
+        elif rampart_group and rampart_group.is_containerized:
+            container_controller = random.choice(task.global_instance_groups)
```
`random.choice()` *sigh* a problem we still need to solve .. resource consumption on the controller side. A nicety of our controller playbook is that it's resource requirements are fixed; unlike customer playbooks that can vary widely. When we go to solve this problem, maybe it's as simple as looking at the current workload in flight (i.e. pending, waiting, running) and just submit the job to the controller with the lowest count. However, that controller _can_ be executing non-isolated workloads. So maybe a combination of looking at the iso queue depth + the already existing mechanism we have for capturing user workload, capacity.

shanemcd replied:
yeeep

chrismeyersfsu replied:
Noodling over the usage of `global_instance_groups` here. This requires users to make sure the special `tower` group exists and to add nodes to in order for this to succeed.  But can we do better than this?

We know that `rampart_group` is a containerized group (it has a kubernetes type cred assigned to it). So should all the `instance`(s) belonging to that special group be considered controllers for that group and we chose from that list?

chrismeyersfsu commented on awx/main/scheduler/task_manager.py:
```diff
@@ -250,6 +254,11 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
+                task.instance_group = rampart_group
+                pod_manager = PodManager(task)
+                pod_manager.deploy_pod()
+                task.execution_node = pod_manager.pod_name
```
`execution_node` is the container and the `instance_group`  is the first of the `preferred_instance_groups` which comes from our existing hierarchy of which instance group to choose logic :+1:

chrismeyersfsu replied:
It's a shame we don't capture the controller node.

AlanCoding replied:
We don't?

https://github.com/ansible/awx/blob/5d6916f69ee1ce39676384514fc0cb79cbe3bdaa/awx/main/models/unified_jobs.py#L586-L590

chrismeyersfsu commented on awx/main/tasks.py:
```diff
@@ -1263,6 +1265,7 @@ def run(self, pk, **kwargs):
                     params.pop('inventory'),
                     os.path.join(private_data_dir, 'inventory')
                 )
+                params['envvars'].pop('HOME')
```
Good find. But would this hurt/break non-container isolated runs? It may be subtle, but users may depend on this dir existing in their playbooks. For example, I had `ansible-galaxy` break on me because it tries to create `~/.ansible_galaxy` but `HOME` wasn't defined so it tried to create in `/`, got permission denied, and ultimately failed my playbook.

shanemcd replied:
Haha. This one was a fun one to track down. This is the wrong `HOME` altogether! This is passed through to the environment on the isolated node but it's actually read / set from the environment on the *controller*.

chrismeyersfsu replied:
We enforce the same dir/home structure on the iso nodes as on non-iso full-tower nodes. Thus, we know `/var/lib/awx` to be the home dir on an ISO node.

So I would say, it is fine to strip this when the iso node is this new container, but leave `HOME` defined for legacy isolated.

chrismeyersfsu commented on awx/main/tasks.py:
```diff
@@ -1598,6 +1601,11 @@ def pre_run_hook(self, job):
                 pu_ig = pu_ig.controller
                 pu_en = settings.CLUSTER_HOST_ID
+
+            if job.instance_group.is_containerized is True:
```
Comparison of is True here may let the programmer think that the corollary of `is False` can be used. However, in `is_containerized` current implementation `None` can be returned when the objective is to convey `False`. I'd say either cast `is_containerized` to bool (my preferred solution) or change this to drop the `is True`

chrismeyersfsu commented on awx/main/tests/functional/task_management/test_container_groups.py:
```diff
@@ -0,0 +1,54 @@
+
+
+@pytest.mark.django_db
+def test_containerized_job(containerized_job):
```
:+1:

chrismeyersfsu commented on awx/main/models/credential/__init__.py:
```diff
@@ -1093,6 +1097,38 @@ def create(self):
 )
 
 
+ManagedCredentialType(
```
We are introducing a new credential type but only making it useable when attached to an instance group. I would imagine this could be useful to customers outside of this context, in their job runs. Should the logic that lives in `kubectl_or_oc_vars` be moved to our general credential type injection so this cred can be re-used? @ryanpetrello @AlanCoding opinions?

AlanCoding replied:
If it cannot be reused, then we should add this to the credential validation such that the "kubernetes" kind cannot be attached to inventory syncs or job runs. There is not a great standard pattern for this.

I was actually very excited about implementing either the [openshift](https://docs.ansible.com/ansible/latest/plugins/inventory/openshift.html) or [k8s](https://docs.ansible.com/ansible/latest/plugins/inventory/k8s.html) inventory plugin, and of course this would use the official credential type. I have some history on that. I was really trying to avoid writing the kubeconfig file, because the plugin itself should be able to process the token. Link https://github.com/openshift/openshift-restclient-python/issues/249

I do need to go back through see how this lays down the credential information and compare to what we might do with module / inventory plugin injection.

jakemcdermott replied:
@chrismeyersfsu how would one go about attaching this credential to jobs (job templates?) in the new scenarios you're speaking of? Are we talking about adding another selectable credential kind to job templates?

chrismeyersfsu replied:
Would fit into our existing cloud credential(s).

chrismeyersfsu commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,86 @@
+
+    @property
+    def namespace(self):
+        return self.pod_definition['metadata']['namespace']
```
See concern about user defined pod definition in `_pod_definition`

chrismeyersfsu commented on awx/playbooks/check_isolated.yml:
```diff
@@ -20,12 +20,48 @@
+      when: ansible_kubectl_host is not defined
+
+    - name: Copy artifacts from the isolated host.
+      synchronize:
```
Although this output doesn't get _directly_ captured, it can end up in log files; we should be good citizens and include ansible `no_log: True`.

AlexSCorey committed:
Add Container Group Form Render and cred type modal pops up

modal render proper cred

WIP

modal rendering properly

card for edit CG renders, no fields

add code mirror and some list styling

address PR issues

shanemcd commented on awx/main/models/credential/__init__.py:
```diff
@@ -1093,6 +1097,38 @@ def create(self):
+            'help_text': ugettext_noop('The OpenShift or Kubernetes API Endpoint to authenticate with.')
+        },{
+            'id': 'bearer_token',
+            'label': ugettext_noop('API authentication bearer token.'),
```
Note to self: add more help text

shanemcd replied:
I forget why I made this note and no one has said anything. Guess it doesn't matter.

shanemcd committed:
Create separate Make target for cleaning API-related artifacts

My workflow for running tests is now:

```
$ docker exec -ti tools_awx_1 make clean-api awx-link test
```

shanemcd committed:
Fix InstanceGroup summary fields

jakemcdermott committed:
link to container group from job runs

jakemcdermott committed:
add is_containerized to ig serializer

AlexSCorey committed:
Adds Tabs to CGs

AlexSCorey committed:
add default pod spec to edit

AlexSCorey committed:
removes instances tab from CGs

AlexSCorey committed:
Show override toggle as off when pod_spec matches default

Commit:
add k8s cred type to awxkit

shanemcd commented:
recheck

kdelee commented:
Current Questions/Observations:

1) is this meant to close https://github.com/ansible/awx/issues/1485 ?
1)  Launching adhoc command on inventory in org with only container group seems to hang. Seeing this looping in logs (been several min. now):
```
2019-09-26 19:57:32,164 DEBUG    awx.main.dispatch delivered b1c15ecf-dd37-4096-a47e-859d35c405fe to worker[592] qsize 0
2019-09-26 19:57:32,169 DEBUG    awx.main.dispatch task b1c15ecf-dd37-4096-a47e-859d35c405fe starting awx.main.scheduler.tasks.run_task_manager(*[])
2019-09-26 19:57:32,172 DEBUG    awx.main.scheduler Running Tower task manager.
2019-09-26 19:57:32,194 DEBUG    awx.main.scheduler Starting Scheduler
2019-09-26 19:57:32,245 DEBUG    awx.main.scheduler Skipping group ContainerGroup - ozet2q0tum, remaining_capacity 0 <= 0
2019-09-26 19:57:32,246 DEBUG    awx.main.scheduler ad_hoc_command 37 (pending) couldn't be scheduled on graph, waiting for next cycle
2019-09-26 19:57:32,247 DEBUG    awx.main.dispatch task b1c15ecf-dd37-4096-a47e-859d35c405fe is finished
```
3) Launching inventory update on inventory in org with only container group also seems to hang indefinitely
loop in logs:
```
 2019-09-26 22:15:33,847 DEBUG    awx.main.dispatch delivered 2589d690-51dc-4f54-8de2-98aede168e4e to worker[592] qsize 0
2019-09-26 22:15:33,849 DEBUG    awx.main.dispatch task 2589d690-51dc-4f54-8de2-98aede168e4e starting awx.main.scheduler.tasks.run_task_manager(*[])
2019-09-26 22:15:33,852 DEBUG    awx.main.scheduler Running Tower task manager.
2019-09-26 22:15:33,872 DEBUG    awx.main.scheduler Starting Scheduler
2019-09-26 22:15:33,952 DEBUG    awx.main.scheduler Skipping group ContainerGroup - ugnbfqlrus, remaining_capacity 0 <= 0
2019-09-26 22:15:33,953 DEBUG    awx.main.scheduler inventory_update 38 (pending) couldn't be scheduled on graph, waiting for next cycle
2019-09-26 22:15:33,955 DEBUG    awx.main.dispatch task 2589d690-51dc-4f54-8de2-98aede168e4e is finished
2019-09-26 22:15:53,823 DEBUG    awx.main.dispatch publish awx.main.scheduler.tasks.run_task_manager(7fd8564e-16e4-42d4-971a-f6e43ffca3d9, queue=awx_private_queue)
```
4) launching project update on project in org with only container group also hanging:
```
2019-09-26 23:18:14,745 DEBUG    awx.main.dispatch delivered 2247ef30-27dd-41ec-9290-4ed551447f6d to worker[592] qsize 0                                                                                                                     
2019-09-26 23:18:14,749 DEBUG    awx.main.dispatch task 2247ef30-27dd-41ec-9290-4ed551447f6d starting awx.main.scheduler.tasks.run_task_manager(*[])                                                                                         
2019-09-26 23:18:14,752 DEBUG    awx.main.scheduler Running Tower task manager.
2019-09-26 23:18:14,786 DEBUG    awx.main.scheduler Starting Scheduler
2019-09-26 23:18:14,837 DEBUG    awx.main.scheduler Skipping group ContainerGroup - cmd5ahmqmp, remaining_capacity 0 <= 0                                                                                                                    
2019-09-26 23:18:14,838 DEBUG    awx.main.scheduler project_update 52 (pending) couldn't be scheduled on graph, waiting for next cycle                                                                                                       
2019-09-26 23:18:14,839 DEBUG    awx.main.dispatch task 2247ef30-27dd-41ec-9290-4ed551447f6d is finished
```
5) can we get the default pod_spec in the OPTIONS to be in JSON instead of YAML?

shanemcd committed:
Change default pod spec in OPTIONS request to json

AlexSCorey committed:
Fixes Credential Type Issue, ExtraVars Toggle Issue, Job Results Inert Link

kdelee commented:
@shanemcd based on conversation with @one-t and @unlikelyzero we would prefer to see job state be `error` when awx is able to bring up container, but unable to connect to it. (Think: bad image that does not have rsync). E.g. we'd like to reserve `failed` for when the user's playbook has a failed task and `error` for when we can't run the job properly for some reason.

kdelee commented:
I'm noticing that for jobs that fail with an error where you cannot connect to the k8s cluster, not seeing container group or execution node linked at all:
![Screenshot from 2019-09-30 15-41-44](https://user-images.githubusercontent.com/29239279/65910578-fd8e3400-e398-11e9-8dd5-86b0b1b2d2be.png)
This is because API does not have this info set at this point of the job launch.
@shanemcd @ryanpetrello is this something y'all think we can get fixed in this PR or should I file a bug for later consideration?

ryanpetrello commented on awx/main/isolated/manager.py:
```diff
@@ -153,6 +169,12 @@ def dispatch(self, playbook=None, module=None, module_args=None):
                                                   extravars=extravars)
+
+        # Do we want to do this for all jobs rather that just containerized jobs?
+        if runner_obj.status == 'failed' and self.instance.is_containerized:
```
This is a good idea, and I think we should do it regardless of whether it's containerized or not.

ryanpetrello commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,150 @@
+
+    fd, path = tempfile.mkstemp(prefix='kubeconfig')
+    with open(path, 'wb') as temp:
+        temp.write(yaml.dump(config).encode())
```
If this contains the raw token value, we should e.g.,

`os.chmod(temp, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)`

shanemcd replied:
Done

ryanpetrello commented on awx/main/scheduler/kubernetes.py:
```diff
@@ -0,0 +1,150 @@
+
+        # We don't do any fancy timeout logic here because it is handled
+        # at a higher level in the job spawning process. See
+        # settings.AWX_ISOLATED_LAUNCH_TIMEOUT and settings.AWX_ISOLATED_CONNECTION_TIMEOUT
```
Is this statement actually true?  This code is called from the `def run` in `tasks.py`, and it seems to me that if the pod status got stuck in pending forever, this code would just happily hang out in this loop until the end of time (which would probably be bad).

ryanpetrello replied:
Maybe we should put some timer code into the `while True:` loop just to be safe e.g.,

```python
start = time.time()
while True:
    if (time.time() - start) > AWX_CONTAINER_GROUP_DEFAULT_LAUNCH_TIMEOUT:
        raise RuntimeError(...)
```

shanemcd replied:
Hrm, I wonder if I wrote this when it was called inside of the task manager.

@ryanpetrello since I'm going to move this into a task, is there a more generic way of handling a timeout when the worker works on stuff?

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1600,6 +1613,8 @@ def should_use_proot(self, job):
         Return whether this task should use proot.
         '''
+        if job.is_containerized:
+            return False # TODO: Get smarter
```
Is this `TODO` still needed?  It seems reasonable to me that containerized jobs would just skip bwrap entirely.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1720,8 +1736,13 @@ def final_run_hook(self, job, status, private_data_dir, fact_modification_times,
+
+        if job.is_containerized:
+            from awx.main.scheduler.kubernetes import PodManager # prevent circular import
+            PodManager(job).delete()
```
We should add some exception handling around this.

```python
try:
    ...
except Exception:
    logger.exception('Failed to clean up pod.')
```

shanemcd replied:
Not ignoring this, but will probably leave for a more comprehensive PR focused on logging. Definitely on my list.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -2537,6 +2558,8 @@ def should_use_proot(self, ad_hoc_command):
         Return whether this task should use proot.
         '''
+        if ad_hoc_command.is_containerized:
+            return False # TODO: Get smarter
```
Same question here as above.

ryanpetrello reviewed (commented):
I left some minor review feedback.  This was a huge undertaking that touched a *lot* of Tower internals - great work on this 😄.

mabashian reviewed (approved):
UI code lgtm

AlexSCorey committed:
IG List responsiveness

wenottingham commented on awx/ui/client/src/instance-groups/instance-groups.strings.js:
```diff
@@ -1,13 +1,16 @@
         INSTANCES_BREADCRUMB_LABEL: t.s('INSTANCES'),
-        ADD_BREADCRUMB_LABEL: t.s('CREATE INSTANCE GROUP')
+        ADD_BREADCRUMB_LABEL: t.s('CREATE INSTANCE GROUP'),
+        ADD_CONTAINER_GROUP_BREADCRUMB_LABEL: t.s('CREATE CONTAINER GROUP')
```
so maybe ('CREATE CONTAINER GROUP (BETA)') here

wenottingham commented on awx/ui/client/src/instance-groups/instance-groups.strings.js:
```diff
@@ -62,6 +66,15 @@ function InstanceGroupsStrings (BaseString) {
         MISSING_PARAMETER: t.s('Instance Group parameter is missing.'),
     };
+    ns.container = {
+        PANEL_TITLE: t.s('Add Container Group'),
```
similarly (BETA) here

AlexSCorey committed:
Add Tech Preview notice to Container Group UI

and some refactoring

shanemcd commented on awx/ui/client/src/instance-groups/main.js:
```diff
@@ -207,11 +385,11 @@ function InstanceGroupsRun ($stateExtender, strings) {
             resolvedModels: InstanceGroupsResolve,
             Dataset: ['GetBasePath', 'QuerySet', '$stateParams',
-                function(GetBasePath, qs, $stateParams) {
+                function (GetBasePath, qs, $stateParams) {
```
@jakemcdermott @AlexSCorey is this valid js?

AlexSCorey replied:
@shanemcd Yes it seems like t just added a space there.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1272,10 +1273,22 @@ def run(self, pk, **kwargs):
                     params.pop('inventory'),
                     os.path.join(private_data_dir, 'inventory')
                 )
+                pod_manager = None
```
@shanemcd what if we moved this _entire_ block up ^^^

ryanpetrello replied:
..to the *very* beginning of the method *before* we do `status = running`:

https://github.com/ansible/awx/pull/4189/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R1146

We could wrap it with a `try...except...` and if we encountered an exception (like a 403 for resource quota problems), we could log it, and then just return (and leave the job in pending state).

ryanpetrello replied:
e.g.,

```python
def run(...):
    if containerized:
        try:
            launch_k8s()
        except 403:
            logger.warning('out of quota')
            return # job stays in pending and the task manager tries again soon
        except Exception:
            self.update_model(pk, status='error')  # other k8s failures should set the status to error
            logger.exception('k8s error happened')
            return
```

ryanpetrello replied:
cc @kdelee

ryanpetrello replied:
`except 403`  <--- obviously pseudo-code

I dunno what actual exception you all were encountering, might need to ask @one-t so we can reproduce and catch it.

ryanpetrello replied:
@shanemcd also, to clarify further, I don't think we should do *any* of this in *this* PR (it's my suggestion for https://github.com/ansible/awx/issues/4910)

shanemcd replied:
We have a patch that solves this but are holding in until after initial merge so we can run it through the full regression suite.

ryanpetrello reviewed (approved):
@shanemcd also, to clarify further, I don't think we should do *any* of https://github.com/ansible/awx/pull/4189/files#r331476690 in *this* PR (it's my suggestion for https://github.com/ansible/awx/issues/4910)

I've got some minor code cleanup comments, but in my opinion, once @kdelee is happy, we should merge what we have here and keep moving.

ryanpetrello commented on awx/main/models/ha.py:
```diff
@@ -92,6 +92,10 @@ class Instance(HasPolicyEditsMixin, BaseModel):
         default=0,
         editable=False,
     )
+    is_containerized = models.BooleanField(
```
Do we need an actual field for this now, since under the current model, it's *always* going to be `False` (i.e., there's no such thing _today_ as an `Instance` object that actually represents a pod).

Seems like we could just lose this field and know that `Instance` objects _always_ represent an actual VM/machine (not a pod), and that containerized instance groups have zero capacity and no "instances"

shanemcd replied:
Good catch, I added this and never went anywhere with it.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -4685,10 +4685,10 @@ class InstanceSerializer(BaseSerializer):
     class Meta:
         model = Instance
-        read_only_fields = ('uuid', 'hostname', 'version')
+        read_only_fields = ('uuid', 'hostname', 'version', 'is_containerized')
```
If you remove https://github.com/ansible/awx/pull/4189/files#r331524755, you can lose this, too.

ryanpetrello commented on awx/main/migrations/0095_v360_container_groups.py:
```diff
@@ -0,0 +1,43 @@
+    ]
+
+    operations = [
+        migrations.AddField(
```
If you remove https://github.com/ansible/awx/pull/4189/files#r331524755, you can lose this, too.

ryanpetrello commented on awx/main/models/ha.py:
```diff
@@ -271,23 +291,27 @@ def schedule_policy_task():
 def on_instance_saved(sender, instance, created=False, raw=False, **kwargs):
     if created or instance.has_policy_changes():
-        schedule_policy_task()
+        if not instance.is_containerized:
```
If you remove https://github.com/ansible/awx/pull/4189/files#r331524755, you can lose this, too.

ryanpetrello commented on awx/main/models/ha.py:
```diff
@@ -271,23 +291,27 @@ def schedule_policy_task():
 @receiver(post_delete, sender=Instance)
 def on_instance_deleted(sender, instance, using, **kwargs):
-    schedule_policy_task()
+    if not instance.is_containerized:
```
If you remove https://github.com/ansible/awx/pull/4189/files#r331524755, you can lose this, too.

ryanpetrello commented:
regate
